### DISCUSSION TITLE
Update config file format for the Agglayer

### DISCUSSION THREAD

**Discussion by Freyskeyd** [agglayer/agglayer#147] [Ideas]
## Rationale for Discussion

*original ticket #83*

The discussion around the configuration format for `agglayer` has arisen due to the need for greater consistency and readability in configuration file. Currently, there is a mix of `PascalCase` and `lowercase` formats, leading to confusion and maintenance challenges in the future. As we'll for sure add more configuration option to allow a wide range of system configuration but also behaviors configuration, we need to think about the best format/shape to offer.

## Format Options
For `agglayer` configuration files, four main formats are being considered:

- `PascalCase`: Often used for class names in object-oriented programming. Examples: `LayerConfig`, `MaxConnections`.

- `lowercase`: A simple format, all in lowercase without spaces or hyphens. Examples: `layerconfig`, `maxconnections`.

- `snake_case`: Uses underscores to separate words, commonly seen in Python. Examples: `layer_config`, `max_connections`.

- `kebab-case`: Uses hyphens to separate words, making it easy to read and type. Examples: `layer-config`, `max-connections`.

### PascalCase

- **Pros**: Common in programming, good readability for developers.
- **Cons**: Can be confused with class names, less intuitive for textual configurations.
```
[Auth.GcpKms]
ProjectId = ""
Location = ""
Keyring = ""
KeyName = ""

[Outbound.Rpc.Settle]
MaxRetries = 10
RetryInterval = 1
Confirmations = 3
```

### lowercase

- **Pros**: Simple, no special characters.
- **Cons**: Difficult to read for complex configurations, lacks visual word separation.

```
[auth.gcpkms]
projectid = ""
location = ""
keyring = ""
keyname = ""

[outbound.rpc.settle]
maxretries = 10
retryinterval = 1
confirmations = 3
```

### snake_case:

- **Pros**: Clear word separation, familiar to many developers.
- **Cons**: Underscores can be harder to type quickly and may be less aesthetically pleasing.

```
# snake_case

[auth.gcp_kms]
project_id = ""
location = ""
keyring = ""
key_name = ""

[outbound.rpc.settle]
max_retries = 10
retry_interval = 1
confirmations = 3
```

### kebab-case:

- **Pros**
	- **Readability**: Hyphens provide clear word separation, enhancing readability.
	- **Uniformity**: Promotes consistency throughout the configuration file.
	- **Widespread Use**: Increasingly popular for configuration files in modern frameworks like YAML and CSS practices.
- **Cons**: May be perceived as less traditional by developers accustomed to other formats.

```
# kebab-case

[auth.gcp-kms]
project-id = ""
location = ""
keyring = ""
key-name = ""

[outbound.rpc.settle]
max-retries = 10
retry-interval = 1
confirmations = 3
```
## Personal point of view

For me, the `kebab-case` format emerges as the best solution for `agglayer` configuration files. As its adoption ensures better readability and increased consistency while facilitating the maintenance.

**Comment by iljakuklic** [comment]
Aesthetically I like kebab case the best.

Wondering though how important is it to maintain backwards compatibility here?

**Reply by Freyskeyd** [reply]
Currently, we have backward compatibility with the old configuration from Go, but I think it can be fairly easy to add/update things without breaking based on what we did so far.

**Reply by muursh** [reply]
I'd vote kebab too tbh.

**Comment by Freyskeyd** [comment]
The format we choose is `kebab-case`.
